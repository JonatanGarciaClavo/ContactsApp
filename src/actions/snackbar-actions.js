import { CLOSE_ERROR_MESSAGE, SET_ERROR_MESSAGE } from '../constants/snackbar-actions-constants';
import _ from 'lodash';

export default {
  closeSnackbar() {
    return dispatch => dispatch({
      type: CLOSE_ERROR_MESSAGE,
    });
  },
  displayError(error) {
    let message = 'Unknown error';
    if (error.message) {
      message = error.message;
    } else if (_.isString(error)) {
      message = error;
    }
    return {
      type: SET_ERROR_MESSAGE,
      message,
    };
  },
};
