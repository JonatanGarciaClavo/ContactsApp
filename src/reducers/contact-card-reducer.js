import Immutable from 'immutable';
import { REQUEST_CONTACT_CARD, REQUEST_DELETE_CONTACT_CARD, REQUEST_CONTACT_CARD_SUCCESS,
  RESET_CONTACT_CARD } from '../constants/contact-card-actions-constants';
import { SET_ERROR_MESSAGE } from '../constants/snackbar-actions-constants';

export const initialState = new Immutable.Map({
  loading: false,
  contact: new Immutable.Map(),
});

const contactCard = (state = initialState, action) => {
  switch (action.type) {
    case (REQUEST_CONTACT_CARD):
    case (REQUEST_DELETE_CONTACT_CARD):
      return state.set('loading', true);
    case (SET_ERROR_MESSAGE):
      return state.set('loading', false);
    case (REQUEST_CONTACT_CARD_SUCCESS):
      return state
        .set('contact', action.contact)
        .set('loading', false);
    case (RESET_CONTACT_CARD):
      return initialState;
    default:
      return state;
  }
};

export default contactCard;
