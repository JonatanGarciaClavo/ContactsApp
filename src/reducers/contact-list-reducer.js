import { Record, List } from 'immutable';
import { REQUEST_CONTACT_LIST, REQUEST_CONTACT_LIST_SUCCESS, CHANGE_LIST_MODE, RESET_CONTACT_LIST }
  from '../constants/contact-list-actions-constants';
import { SET_ERROR_MESSAGE } from '../constants/snackbar-actions-constants';
import { LIST_MODE } from '../constants/contact-list-mode-constants';

export const ContactList = new Record({
  loading: false,
  contacts: new List(),
  orderBy: 'name',
  mode: LIST_MODE,
});

const contactList = (state = new ContactList(), action) => {
  switch (action.type) {
    case (REQUEST_CONTACT_LIST):
      return state.set('loading', true);
    case (SET_ERROR_MESSAGE):
      return state.set('loading', false);
    case (REQUEST_CONTACT_LIST_SUCCESS):
      return state
        .set('contacts', action.contacts)
        .set('loading', false);
    case (CHANGE_LIST_MODE):
      if (action.mode === state.get('mode')) {
        return state;
      }
      return state.set('mode', action.mode || LIST_MODE);
    case (RESET_CONTACT_LIST):
      return new ContactList();
    default:
      return state;
  }
};

export default contactList;
