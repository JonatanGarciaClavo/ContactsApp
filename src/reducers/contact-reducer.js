import _ from 'lodash';
import { Record, Map, fromJS } from 'immutable';
import { REQUEST_CONTACT, REQUEST_SAVE_CONTACT, REQUEST_CONTACT_SUCCESS, UPDATE_CONTACT_ATTRIBUTE,
  VALIDATE_CONTACT, VALIDATE_CONTACT_ATTRIBUTE, RESET_CONTACT,
  } from '../constants/contact-actions-constants';
import { SET_ERROR_MESSAGE } from '../constants/snackbar-actions-constants';
import ContactModel from '../models/ContactModel';

const requiredValidation = value =>
  value !== undefined && value !== null && (!_.isEmpty(value) || _.isNumber(value));

const isEmail = (value) => {
  if (value) {
    const regExp = /^((([a-z]|\d|[!#\$%&'\*\+\-\/=\?\^_`{\|}~]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])+(\.([a-z]|\d|[!#\$%&'\*\+\-\/=\?\^_`{\|}~]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])+)*)|((\x22)((((\x20|\x09)*(\x0d\x0a))?(\x20|\x09)+)?(([\x01-\x08\x0b\x0c\x0e-\x1f\x7f]|\x21|[\x23-\x5b]|[\x5d-\x7e]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(\\([\x01-\x09\x0b\x0c\x0d-\x7f]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]))))*(((\x20|\x09)*(\x0d\x0a))?(\x20|\x09)+)?(\x22)))@((([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))\.)+(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))$/i;// eslint-disable-line
    return regExp.test(value);
  }
  return true;
};

const isUrl = (value) => {
  if (value) {
    const regExp = new RegExp("^(?:(?:https?|ftp):\/\/)(?:\S+(?::\S*)?@)?(?:(?!10(?:\.\d{1,3}){3})(?!127(?:\.\d{1,3}){3})(?!169\.254(?:\.\d{1,3}){2})(?!192\.168(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z\u00a1-\uffff0-9]+-?)*[a-z\u00a1-\uffff0-9]+)(?:\.(?:[a-z\u00a1-\uffff0-9]+-?)*[a-z\u00a1-\uffff0-9]+)*(?:\.(?:[a-z\u00a1-\uffff]{2,})))(?::\d{2,5})?(?:\/[^\s]*)?");// eslint-disable-line
    return regExp.test(value);
  }
  return true;
};

const applyValidator = (ruleObject, value) => {
  let error;
  let functionValidator;
  if (_.isFunction(ruleObject.rule)) {
    functionValidator = ruleObject.rule;
  }
  if (functionValidator && !functionValidator(value)) {
    error = ruleObject.error || `${ruleObject.rule} not followed`;
  } else if (!functionValidator) {
    console.warn('applyValidator: %s does not exists', ruleObject.rule);// eslint-disable-line
  }
  return error;
};

const validateAll = (rulesObject, contact) => {
  const errors = {};

  _.each(rulesObject, (rules, name) => {
    if (rules && _.isArray(rules)) {
      const value = contact[name];
      _.each(rules, (ruleObject) => {
        const error = applyValidator(ruleObject, value);
        if (error) {
          errors[name] = error;
        }
      });
    } else {
      console.warn('validateAll: missing parameters rules ');// eslint-disable-line
    }
  });

  return _.isEmpty(errors) ? false : errors;
};

const rulesObject = {
  name: [{ rule: requiredValidation, error: 'Name is required' }],
  email: [
    { rule: requiredValidation, error: 'Email is required' },
    { rule: isEmail, error: 'Email format is incorrect' },
  ],
  imgUrl: [{ rule: isUrl, error: 'Invalid url' }],
};

export const Contact = new Record({
  loading: false,
  isModified: false,
  contact: new ContactModel(),
  errors: new Map(),
});


const contact = (state = new Contact(), action) => {
  const { name, value } = action;
  let error;
  let errors;
  switch (action.type) {
    case (REQUEST_SAVE_CONTACT):
    case (REQUEST_CONTACT):
      return state.set('loading', true);
    case (SET_ERROR_MESSAGE):
      return state.set('loading', false);
    case (REQUEST_CONTACT_SUCCESS):
      return state
        .set('contact', action.contact || new ContactModel())
        .set('loading', false)
        .set('errors', new Map());
    case (UPDATE_CONTACT_ATTRIBUTE):
      return state
        .set('isModified', true)
        .setIn(['contact', name], value);
    case VALIDATE_CONTACT_ATTRIBUTE:
      error = validateAll(_.pick(rulesObject, name), { [name]: value });
      if (error && error[name]) {
        return state.setIn(['errors', name], error[name]);
      }
      return state.deleteIn(['errors', name]);
    case VALIDATE_CONTACT:
      errors = validateAll(rulesObject, state.get('contact').toJS());
      return state.set('errors', new Map(fromJS(errors || {})));
    case RESET_CONTACT:
      return new Contact();
    default:
      return state;
  }
};

export default contact;
