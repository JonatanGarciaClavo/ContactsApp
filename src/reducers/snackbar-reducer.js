import { CLOSE_ERROR_MESSAGE, SET_ERROR_MESSAGE } from '../constants/snackbar-actions-constants';

export const initialState = {
  open: false,
  message: 'Unknown error',
};

const snackbar = (state = initialState, action) => {
  switch (action.type) {
    case (CLOSE_ERROR_MESSAGE):
      return Object.assign({}, initialState);
    case (SET_ERROR_MESSAGE):
      return Object.assign({}, state, { open: true, message: action.message });
    default:
      return state;
  }
};

export default snackbar;
