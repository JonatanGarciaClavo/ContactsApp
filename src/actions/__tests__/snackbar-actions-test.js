import mockStore from 'redux-mock-store';
import { closeSnackbar, displayError } from '../snackbar-actions';
import { initialState } from '../../reducers/snackbar-reducer';

const store = mockStore({ snackbar: initialState });

describe('Test snackbar actions', () => {
  beforeEach(() => {
    store.clearActions();
  });
  it('should handle closeSnackbar action', () => {
    store.dispatch(closeSnackbar());
    expect(store.getActions()).toMatchSnapshot();
  });
  it('should handle displayError no error action param action', () => {
    store.dispatch(displayError());
    expect(store.getActions()).toMatchSnapshot();
  });
  it('should handle displayError js Error', () => {
    store.dispatch(displayError(new Error('Test error')));
    expect(store.getActions()).toMatchSnapshot();
  });
  it('should handle displayError string error', () => {
    store.dispatch(displayError('Test string error'));
    expect(store.getActions()).toMatchSnapshot();
  });
});
