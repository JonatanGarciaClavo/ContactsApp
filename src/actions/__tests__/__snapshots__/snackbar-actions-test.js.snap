exports[`Test snackbar actions should handle closeSnackbar action 1`] = `
Array [
  Object {
    "type": "CLOSE_ERROR_MESSAGE",
  },
]
`;

exports[`Test snackbar actions should handle displayError js Error 1`] = `
Array [
  Object {
    "message": "Test error",
    "type": "SET_ERROR_MESSAGE",
  },
]
`;

exports[`Test snackbar actions should handle displayError no error action param action 1`] = `
Array [
  Object {
    "message": "Unknown error",
    "type": "SET_ERROR_MESSAGE",
  },
]
`;

exports[`Test snackbar actions should handle displayError string error 1`] = `
Array [
  Object {
    "message": "Test string error",
    "type": "SET_ERROR_MESSAGE",
  },
]
`;
